embassy-stm32/src/dma/ringbuffer.rs

```diff
@@ -54,151 +53,130 @@ pub trait DmaCtrl {
-                }
-            }
+            /*
+                first, check if the dma has wrapped around more than once
 
-            self.first = (self.first + len) % self.dma_buf.len();
-            Ok(len)
+                then, get the current position of of the dma write and check
+                if it's inside data we could have copied
+            */
+            let pos = self.pos(dma.get_remaining_transfers());
+            if dma.get_complete_count() > 1 || pos > self.start || pos < end {
+                Err(OverrunError)
+            } else {
+                self.start = (self.start + len) % self.len();
+
+                Ok((len, self.start + end))
+            }
         } else {
             // The available, unread portion in the ring buffer DOES wrap
             // The DMA writer has wrapped since we last read and is currently
             // writing (or the next byte added will be) in the beginning of the ring buffer.
 
-            let complete_count = dma.get_complete_count();
-            if complete_count > 1 {
-                return Err(OverrunError);
-            }
-
-            // If the unread portion wraps then the writer must also have wrapped
-            assert!(complete_count == 1);
-
-            if self.first + buf.len() < self.dma_buf.len() {
-                // The provided read buffer is not large enough to include all bytes from the tail of the dma buffer.
+            // The provided read buffer is large enough to include all bytes from the tail of the dma buffer,
+            // so the next read will not have any unread tail bytes in the ring buffer.
 
-                // Copy out from the dma buffer
-                let len = self.copy_to(buf, self.first..self.dma_buf.len());
+            // Copy out from the dma buffer
+            let tail = self.copy_to(buf, self.start..self.len());
+            let head = self.copy_to(&mut buf[tail..], 0..end);
 
-                compiler_fence(Ordering::SeqCst);
+            compiler_fence(Ordering::SeqCst);
 
-                // We have now copied out the data from dma_buf
-                // Make sure that the just read part was not overwritten during the copy
-                self.ndtr = dma.ndtr();
-                if self.end() > self.first || dma.get_complete_count() > 1 {
-                    // The writer has entered the data that we have just read since we read out `end` in the beginning and until now.
-                    return Err(OverrunError);
-                }
+            /*
+                first, check if the dma has wrapped around more than once
 
-                self.first = (self.first + len) % self.dma_buf.len();
-                Ok(len)
+                then, get the current position of of the dma write and check
+                if it's inside data we could have copied
+            */
+            let pos = self.pos(dma.get_remaining_transfers());
+            if dma.reset_complete_count() > 1 || pos > self.start || pos < end {
```

Comment:
Maybe this should be `dma.reset_complete_count() > 1 || pos > self.start || dma.reset_complete_count() > 0` ?

Reply:
The benefit of that condition would be marginal, in my opinion.